lib/Modules/ComputeHooks/ComputeHooks.psm1

```diff
@@ -108,972 +96,437 @@ function Install-Prerequisites {
-    $vhost = Get-JujuCharmConfig -Scope 'rabbit-vhost'
-    if (!$username -or !$vhost){
-        Write-JujuError "Missing required charm config options: rabbit-user or rabbit-vhost"
+function Get-DataPort {
+    # try and set up bonding early. This will create
+    # a new Net-LbfoTeam and try to acquire an IP address
+    # via DHCP. This interface may receive os-data-network IP.
+    $bondName = New-BondInterface
+    $managementOS = Get-JujuCharmConfig -Scope "vmswitch-management"
+
+    $netType = Get-NetType
+    if ($netType -eq "ovs") {
+        Write-JujuInfo "Trying to fetch OVS data port"
+        $dataPort = Get-OVSDataPort
+        return @($dataPort[0], $false)
     }
 
-    $required = @{
-        "hostname"=$null;
-        "password"=$null;
+    if ($bondName) {
+        $adapter = Get-NetAdapter -Name $bondName
+        return @($adapter, $managementOS)
     }
 
-    $optional = @{
-        "vhost"=$null;
-        "username"=$null;
-        "ha_queues"=$null;
+    Write-JujuInfo "Trying to fetch data port from config"
+    $nic = Get-InterfaceFromConfig
+    if (!$nic) {
+        $nic = Get-FallbackNetadapter
+        $managementOS = $true
     }
+    $nic = Get-RealInterface $nic[0]
+    return @($nic[0], $managementOS)
+}
 
-    $ctx = Get-JujuRelationContext -Relation "amqp" -RequiredContext $required -OptionalContext $optional
+function Start-ConfigureVMSwitch {
+    $VMswitchName = Get-JujuVMSwitch
+    $vmswitch = Get-VMSwitch -SwitchType External -Name $VMswitchName -ErrorAction SilentlyContinue
 
-    $data = @{}
+    if ($vmswitch) {
+        return $true
+    }
 
-    if($ctx.Count) {
-        if(!$ctx["username"]) {
-            $data["rabbit_userid"] = $username
-        } else {
-            $data["rabbit_userid"] = $ctx["username"]
-        }
-        if(!$ctx["vhost"]){
-            $data["rabbit_virtual_host"] = $vhost
-        } else {
-            $data["rabbit_virtual_host"] = $ctx["vhost"]
-        }
-        if($ctx["ha_queues"]) {
-            $data["rabbit_ha_queues"] = "True"
-        } else {
-            $data["rabbit_ha_queues"] = "False"
+    $dataPort, $managementOS = Get-DataPort
+    $VMswitches = Get-VMSwitch -SwitchType External -ErrorAction SilentlyContinue
+    if ($VMswitches -and $VMswitches.Count -gt 0) {
+        foreach ($i in $VMswitches) {
+            if ($i.NetAdapterInterfaceDescription -eq $dataPort.InterfaceDescription) {
+                Rename-VMSwitch $i -NewName $VMswitchName
+                Set-VMSwitch -Name $VMswitchName -AllowManagementOS $managementOS
+                return $true
+            }
         }
-        $data["rabbit_host"]=$ctx["hostname"];
-        $data["rabbit_password"]=$ctx["password"];
     }
-    return $data
+
+    Write-JujuInfo "Adding new vmswitch: $VMswitchName"
+    New-VMSwitch -Name $VMswitchName -NetAdapterName $dataPort.Name -AllowManagementOS $managementOS
+    return $true
 }
 
-function Get-NeutronUrl {
+function Start-DownloadFile {
```

Comment:
yes. It's [Invoke-FastWebRequest](https://github.com/cloudbase/juju-powershell-modules/blob/master/JujuHelper/JujuHelper.psm1#L83-L178), but the method here it's just a wrapper. It accepts a url of the form `<url>#md5=<checksum>`, parses it, checks if the destination exists and if so and it matches the checksum, return the path for it, otherwise download the file.